lib/utils/typecast_utils.dart

```diff
@@ -36,4 +37,63 @@ class TypeCast {
 
     return null;
   }
+
+  /// Checks for nullable type casts
+  /// Only one case `Type? is Type` always valid assertion case.
+  bool isNullableCompatibility() {
+    final isObjectTypeNullable =
+        source.nullabilitySuffix != NullabilitySuffix.none;
+    final isCastedTypeNullable =
+        target.nullabilitySuffix != NullabilitySuffix.none;
+
+    return isObjectTypeNullable && !isCastedTypeNullable;
+  }
+
+  /// Checks that type checking is unnecessary
+  /// [source] is the source expression type
+  /// [target] is the type against which the expression type is compared
+  /// [isReversed] true for opposite comparison, i.e 'is!'
+  /// and false for positive comparison, i.e. 'is', 'as' or 'whereType'
+  bool isUnnecessaryTypeCheck({
+    bool isReversed = false,
+  }) {
```

Comment:
isReversed could be a class member?
```suggestion
  bool get isUnnecessaryTypeCheck {
```

Reply:
It can be, for the expense of creating on additional instance of TypeCast in avoid_unnecessary_type_assertions_rule. Should be fine